nixos/modules/services/matrix/mjolnir.nix

```
@@ -0,0 +1,243 @@
+{ config, lib, pkgs, ... }:
+
+with lib;
+let
+  cfg = config.services.mjolnir;
+
+  yamlConfig = {
+    inherit (cfg) dataPath managementRoom protectedRooms;
+
+    accessToken = "@ACCESS_TOKEN@"; # will be replaced in "generateConfig"
+    homeserverUrl =
+      if cfg.pantalaimon.enable then
+        "http://${cfg.pantalaimon.options.listenAddress}:${toString cfg.pantalaimon.options.listenPort}"
+      else
+        cfg.homeserverUrl;
+
+    pantalaimon = {
+      inherit (cfg.pantalaimon) username;
+
+      use = cfg.pantalaimon.enable;
+      password = "@PANTALAIMON_PASSWORD@"; # will be replaced in "generateConfig"
+    };
+  };
+
+  moduleConfigFile = pkgs.writeText "module-config.yaml" (
+    generators.toYAML { } (filterAttrs (_: v: v != null)
+      (fold recursiveUpdate { } [ yamlConfig cfg.settings ])));
+
+  # these config files will be merged one after the other to build the final config
+  configFiles = [
+    "${pkgs.mjolnir}/share/mjolnir/config/default.yaml"
+    moduleConfigFile
+  ];
+
+  # this will generate the default.yaml file with all configFiles as inputs and
+  # replace all secret strings using replace-secret
+  generateConfig = pkgs.writeShellScript "mjolnir-generate-config" (
+    let
+      yqEvalStr = concatImapStringsSep " * " (pos: _: "select(fileIndex == ${toString (pos - 1)})") configFiles;
+      yqEvalArgs = concatStringsSep " " configFiles;
+    in
+    ''
+      # mjolnir will try to load a config from "./config/default.yaml" in the working directory
+      # -> let's place the generated config there
+      mkdir -p ${cfg.dataPath}/config
+
+      # merge all config files into one, overriding settings of the previous one with the next config
+      # e.g. "eval-all 'select(fileIndex == 0) * select(fileIndex == 1)' filea.yaml fileb.yaml" will merge filea.yaml with fileb.yaml
+      ${pkgs.yq-go}/bin/yq eval-all -P '${yqEvalStr}' ${yqEvalArgs} > ${cfg.dataPath}/config/default.yaml
+
+      ${optionalString (cfg.accessTokenFile != null) ''
+        ${pkgs.replace-secret}/bin/replace-secret '@ACCESS_TOKEN@' '${cfg.accessTokenFile}' ${cfg.dataPath}/config/default.yaml
+      ''}
+      ${optionalString (cfg.pantalaimon.passwordFile != null) ''
+        ${pkgs.replace-secret}/bin/replace-secret '@PANTALAIMON_PASSWORD@' '${cfg.pantalaimon.passwordFile}' ${cfg.dataPath}/config/default.yaml
+      ''}
+    ''
+  );
+in
+{
+  options.services.mjolnir = {
+    enable = mkEnableOption "mjolnir";
+
+    homeserverUrl = mkOption {
+      type = types.str;
+      default = "https://matrix.org";
+      description = ''
+        Where the homeserver is located (client-server URL).
+
+        When using <literal>pantalaimon</literal>, this option will become the homeserver to which <literal>pantalaimon</literal> connects.
+        The listen address of <literal>pantalaimon</literal> will become the <literal>homeserverUrl</literal> of <literal>pantalaimon</literal>.
+      '';
```

I'm not sure what the best way to do this is, but I think it's weird to have this option totally ignored when `pantalaimon` is enabled.

```suggestion
      default = if cfg.pantalaimon.enable then
        "http://${cfg.pantalaimon.options.listenAddress}:${toString cfg.pantalaimon.options.listenPort}"
      else
        "https://matrix.org";
      defaultText = ''
        if <literal>cfg.pantalaimon.enable</literal> is <literal>true</literal>, then the pantalaimon
        listen address and port. Otherwise, <literal>"htps://matrix.org"</literal>
      '';
      description = ''
        Where the homeserver is located (client-server URL).

        When using <literal>pantalaimon</literal>, this option will become the homeserver to which <literal>pantalaimon</literal> connects.
        The listen address of <literal>pantalaimon</literal> will become the <literal>homeserverUrl</literal> of <literal>pantalaimon</literal>.
      '';
```

---

This option isn't ignored when `pantalaimon.enable == true`:

1. If `services.mjolnir.pantalaimon.enable == false`: just use the value as `homeserverUrl` in the mjolnir config. Don't care about pantalaimon.
2. If `services.mjolnir.pantalaimon.enable == true`: "copy" the value to `services.pantalaimon-headless."mjolnir".homeserver` and set `http://<pantalaimonListen>` as the new `homeserverUrl`.

Maybe a bit "magical" - I found a typo in the previous description (pantalaimon -> mjolnir in the last line, last &lt;literal&gt;), maybe now it's clearer.